### PR TITLE
C++: Sign analysis library

### PR DESCRIPTION
**Pull request by rdmarsh2**
This is the first part of a port of the Java range analysis library using the IR. It's based off #197, but I wanted to get it out for a round of review now. deeeb25 is the first commit that isn't part of #197.

I'm expecting the main review to come from @jbj, but I'd also like a look from @dave-bartolomeo for the IR changes and from @aschackmull since he wrote the Java library.

### CODE REVIEW

**Comment by aschackmull** [comment]
Have you tried some tests for completeness on some complex snapshots? I.e. testing that every integral-typed expression has sign information.  I seem to remember needing to close some holes in various corner cases when doing the java version.

**Comment by aschackmull** [comment]
> The Java guards library includes a set of "implies" predicates to handle
short-circuiting conditionals. C++ handles those in IR generation, so
dominance on the IR produces correct results for controlling blocks.

This is not entirely correct.  The java CFG also handles short-circuiting conditionals, so dominance does "the right thing".  The reason why these are also included in the `implies` family of predicates is to support when such conditions occur in non-boolean contexts.  E.g.:
```
boolean b = cond1 && cond2;
if (b) {
  foo();
}
```
In this case we'd like both `cond1` and `cond2` to act as guards for `foo()`, which isn't directly given from dominance.

**Comment by rdmarsh2** [comment]
I'll fix the test failures and run on some real snapshots this afternoon. I'll try and get the implies predicates done tomorrow.

File: cpp/ql/src/semmle/code/cpp/rangeanalysis/SignAnalysis.qll [hunk 1]
```diff
@@ -207,22 +207,20 @@ private predicate unknownSign(Instruction i) {
 }
 
 /**
- * Holds if `lowerbound` is a lower bound for `bounded` at `pos`. This is restricted
+ * Holds if `lowerbound` is a lower bound for `bounded``. This is restricted
  * to only include bounds for which we might determine a sign.
  */
-private predicate lowerBound(IRGuardCondition comp, Instruction lowerbound, Instruction bounded, Instruction pos, boolean isStrict) {
+private predicate lowerBound(IRGuardCondition comp, Operand lowerbound, Operand bounded, boolean isStrict) {
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/rangeanalysis/SignAnalysis.qll]: To port this library to `Operand`, I thought you'd just change pairs of `(Instruction instr, Instruction pos)` into `Operand op`, but this commit also changes occurrences like `Instruction lowerbound` to have type `Operand` without calling its `.getInstruction()`. That will increase the size of each predicate like this by effectively joining with all the places where `lowerbound` is used. Is there a good reason to do that?

File: cpp/ql/src/semmle/code/cpp/rangeanalysis/SignAnalysis.qll [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+  string toString() {
+    result = "-" and this = TNeg() or
+    result = "0" and this = TZero() or
+    result = "+" and this = TPos()
+  }
+  Sign inc() {
+    this = TNeg() and result = TNeg() or
+    this = TNeg() and result = TZero() or
+    this = TZero() and result = TPos() or
+    this = TPos() and result = TPos()
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/rangeanalysis/SignAnalysis.qll]: This code assumes that overflow can't occur, but we often get FP reports for the existing range analysis library because overflow is relied on in practice. See, for example, https://discuss.lgtm.com/t/false-positive-c-comparison-result-is-always-the-same/1428.

I expect that we'll need to model overflow for unsigned integers, but we can continue to ignore it for signed integers because it's undefined there. For a case like `inc`, this means adding that `this = TPos() and result = TZero()` in the unsigned case.

Comment (aschackmull) [reply cpp/ql/src/semmle/code/cpp/rangeanalysis/SignAnalysis.qll]: For the corresponding query in java, FPs for overflow checks are instead excluded in the query.  This works because overflow checks are pretty idiomatic and easy to recognise, and thus allows the rangeanalysis to stay (mostly) oblivious to such overflows.

File: cpp/ql/src/semmle/code/cpp/rangeanalysis/SignAnalysis.qll (hunk 1 above)
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/rangeanalysis/SignAnalysis.qll]: This allows us to be more precise when `lowerbound` has a lower bound, as in the following (contrived) example:
```
int f(int x, int y) {
  if (x < 0) {
    return x;
  }
  iif (x < y) {
   return y; // y is strictly positive because of the bound on x above
  }
  return 0;
}```

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/rangeanalysis/SignAnalysis.qll]: Nice.

Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/rangeanalysis/SignAnalysis.qll]: I've added a test demonstrating this for lower bounds

Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/rangeanalysis/SignAnalysis.qll]: Oh ,there actually is a cartesian product there - I'll see if I can add it in directly or if I need to do a rewrite of `IRGuards.qll` to make it work

File: cpp/ql/src/semmle/code/cpp/rangeanalysis/SignAnalysis.qll (hunk 2 above)
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/rangeanalysis/SignAnalysis.qll]: Okay, let's go with the Java approach for now and evaluate whether to change it later, when we have queries and results to look at.

**Comment by rdmarsh2** [comment]
From testing on LuaJit, it doesn't look like there's a significant impact on `IRGuards` performance; both runs spent approximately 17 seconds evaluating the IRGuards stages.
